### ISSUE TITLE
Plugin script loading order w/ Built-in AF_Packet

### ISSUE THREAD

**Issue by dopheide-esnet** [zeek/zeek#2528]
In testing the new 'master' with build-in AF_Packet, we ran into an issue where we could redef AF_Packet variable in our cluster-layout.zeek:

error in /usr/local/zeek/versions/f8eb2d92414da0e999fb268490dd4e74aac58084/share/zeek/site/cluster-layout.zeek, line 42: "redef" used but not previously defined (AF_Packet::fanout_mode)
internal error in /usr/local/zeek/versions/f8eb2d92414da0e999fb268490dd4e74aac58084/share/zeek/site/cluster-layout.zeek, line 44: Failed to find variable named: AF_Packet::buffer_size

Arne suggested adding @load builtin-plugins/__load__.zeek to make sure the plugin scripts get loaded ahead of time, which worked.   We could also potentially just @load builtin-plugins/Zeek_AF_Packet, but that doesn't solve the general case.

**Comment by dopheide-esnet** [comment]
To be clear, it's just a script load order issue.  Messing with cluster-layout.zeek isn't very common, so the question I guess is whether plugin scripts should be loaded before the cluster is initialized or not.

**Comment by awelzel** [comment]
I had looked a bit moving the builtin plugins `__load__.zeek` to earlier. But with a builtin Spicy plugin, the Spicy `__load__.zeek`  is actually what's going through its default.zeek and loading the notice-framework and then initializing the cluster-framework and then the `cluster-layout.zeek` is loaded. It's messy.

I had chatted a bit with @sethhall  and he said fairly strongly nothing else should be done in `cluster-layout.zeek`, but setting `Cluster::nodes`  - anything else may act surprising. The `@load` I had proposed isn't a great suggestion either.

So even if the redef's work with an external AF_Packet plugin within `cluster-layout.zeek` (and that may be luck depending whether you use an external spicy plugin, too) the easiest fix may be to move the redefs out of `cluster-layout.zeek` and limit it to `Cluster::nodes` modifications.

**Comment by ckreibich** [comment]
This touches a few things that I think have remained nebulous for people. The first one is that there's historically been a tendency to not do `@load`s because "it is known" that such loads already happened elsewhere. You wouldn't/shouldn't do that in other languages either, and there's no cost to loading the same script twice — Zeek internally recognizes this. So such `@loads` are encouraged. Here, though, I guess it's tricky to figure out what to load because built-in plugins are a bit under-documented.

There's also the fact that plugins support two kinds of entry points to script loading, via `__load__.zeek` and `__preload__.zeek`, which can matter depending on context, but shouldn't matter here, but also needs better documentation.

Long story short — I'm going to add the needs-docs tag to this one so we don't lose track.

**Comment by awelzel** [comment]
> The first one is that there's historically been a tendency to not do @loads because "it is known" that such loads already happened elsewhere. You wouldn't/shouldn't do that in other languages either, and there's no cost to loading the same script twice — Zeek internally recognizes this. So such @loads are encouraged. Here, though, I guess it's tricky to figure out what to load because built-in plugins are a bit under-documented.

My understanding (and I had actually run into this myself) is that loading something within `cluster-layout.zeek` may cause trouble and maybe we should document that.

Maybe the docs should put a warning around `cluster-layout.zeek` to really only do the Cluster::nodes part.

**Comment by ckreibich** [comment]
Right, `cluster-layout.zeek` is unusual because it is loaded straight from a framework but we also ask users to provide it. This makes it much more sensitive than scripts the user naturally tucks onto the end of the overall load order, like `local.zeek`. You're right to suggest warning folks about this. I meant it in the broader sense, when writing your packages for example.

**Comment by ckreibich** [comment]
Let me know if I'm overlooking anything here, but I don't think there's a strict requirement for this approach. The cluster framework's `zeek_init()` in `main.zeek` already does a bit of sanity-checking of the `nodes` table; it could just as well check whether the table is defined at all (which could then happen via regular redefs in `local.zeek` or elsewhere) and abort if not. This would require changes to `zeekctl` which has me thinking we shouldn't do this now, but perhaps when we eventually remove it.

**Comment by awelzel** [comment]
> it could just as well check whether the table is defined at all (which could then happen via regular redefs in local.zeek or elsewhere) and abort if not. This would require changes to zeekctl which has me thinking we shouldn't do this now, but perhaps when we eventually remove it.

Yeah, it would be nice to not have the magic `cluster-layout.zeek` file at all.

@ckreibich - for the "docs required" flag you placed on here: Preferably builtin-plugins behave close to how it would act externally (the fact that it did't for AF_Packet was due to redef'ing within `cluster-layout.zeek` which we seem to agree can be brittle - #2534 was merged). Further, with #2520 fixing the `__preload__.zeek` it should mostly hold now.

Is there something todo to fulfill the "docs required" label you put here, or can I just close?

**Comment by ckreibich** [comment]
Let's just close it Arne. I think with #2534 we're okay here.
